hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/AzcopyToolHelper.java

```diff
@@ -0,0 +1,274 @@
+  }
+
+  /**
+   * Create a file with implicit parent in the container using Azcopy tool.
+   * @param absolutePathToBeCreated absolute path to be created.
+   * @throws Exception if file creation fails.
+   */
+  public void createFileUsingAzcopy(String absolutePathToBeCreated) throws Exception {
+    if (absolutePathToBeCreated != null) {
+      absolutePathToBeCreated = absolutePathToBeCreated.replace(
+          ABFS_DFS_DOMAIN_NAME, ABFS_BLOB_DOMAIN_NAME) + sasToken;
+      runShellScript(fileCreationScriptPath, absolutePathToBeCreated);
+    }
+  }
+
+  /**
+   * Create a implicit folder with implicit parent in the container using Azcopy tool.
+   * @param absolutePathToBeCreated absolute path to be created.
+   * @throws Exception
+   */
+  public void createFolderUsingAzcopy(String absolutePathToBeCreated) throws Exception {
+    if (absolutePathToBeCreated != null) {
+      absolutePathToBeCreated = absolutePathToBeCreated.replace(
+          ABFS_DFS_DOMAIN_NAME, ABFS_BLOB_DOMAIN_NAME) + sasToken;
+      runShellScript(folderCreationScriptPath, absolutePathToBeCreated);
+    }
+  }
+
+  private File findHadoopAzureDir() throws FileNotFoundException {
+    // Find the hadoop-azure directory from the current working directory.
+    File hadoopAzureDir;
+    File currentDir = new File(System.getProperty(USER_DIR_SYSTEM_PROPERTY));
+    if (!currentDir.isDirectory() && !currentDir.getName().equals(HADOOP_AZURE_DIR)) {
+      hadoopAzureDir = findHadoopAzureDir(currentDir);
+      if (hadoopAzureDir == null) {
+        throw new FileNotFoundException(HADOOP_AZURE_DIR + DIR_NOT_FOUND_ERROR);
+      }
+    } else {
+      hadoopAzureDir = currentDir;
+    }
+    return hadoopAzureDir;
+  }
+
+  private File findHadoopAzureDir(File dir) {
+    if (dir == null) {
+      return null;
+    }
+    File[] files = dir.listFiles();
+    if (files == null) {
+      return null;
+    }
+    for (File file : files) {
+      if (file.isDirectory() && file.getName().equals(HADOOP_AZURE_DIR)) {
+        return file;
+      } else {
+        File hadoopAzureDir = findHadoopAzureDir(file);
+        if (hadoopAzureDir != null) {
+          return hadoopAzureDir;
+        }
+      }
+    }
+    return null;
+  }
+
+  // Synchronized method to download azcopy executable if not present.
+  private void downloadAzcopyExecutableIfNotPresent()
+      throws IOException, InterruptedException {
+    // Check if azcopy directory is present in the hadoop-azure directory.
+    // If not present, create a directory and stop other process here only.
+    File azcopyDir = new File(azcopyDirPath);
+    if (!azcopyDir.exists()) {
+      azcopyDir.mkdir();
+    } else {
+      return;
+    }
+
+    // Check if azcopy tool is present in the azcopy directory.
+    File azcopyFile = new File(azcopyExecutablePath);
+    if (!azcopyFile.exists()) {
+      // If azcopy tool is not present, download and extract it
+      String[] downloadCmdArr = {AZCOPY_CMD_SHELL,
```

Comment:
what if multiple jvms reach here. Can we synchronize here also. I would suggest one thing, when we are doing mkdir, we also check the result of azcopy.mkdir(), now I believe from https://docs.oracle.com/javase/8/docs/api/java/io/File.html#mkdir-- and assuming the OS's fs would maintain atomicity on mkdir-syscall, that only one of the syscall would return true and other would return false.

Reply:
Now, I was thinking if we are able to synchronize through this, we can remove need of UUID-based shell files.


What you feel about it?

Reply:
Will take this

Reply:
Taken